testgrid/cmd/configurator/prow.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"k8s.io/test-infra/testgrid/config"
+	"path"
+	"strings"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	prowConfig "k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/pod-utils/downwardapi"
+	prowGCS "k8s.io/test-infra/prow/pod-utils/gcs"
+)
+
+const testgridTestGroupAnnotation = "testgrid-test-group"
+const testgridDashboardsAnnotation = "testgrid-dashboards"
+const testgridTabNameAnnotation = "testgrid-tab-name"
+
+func applySingleProwjobAnnotations(c *Config, pc *prowConfig.Config, j prowConfig.JobBase, jobType prowapi.ProwJobType) error {
+	tabName := j.Name
+	testGroupName := j.Name
+
+	testGroupName, mustMakeGroup := j.Annotations[testgridTestGroupAnnotation]
+	dashboards, addToDashboards := j.Annotations[testgridDashboardsAnnotation]
+	mightMakeGroup := mustMakeGroup || addToDashboards
+
+	if mightMakeGroup {
+		if testGroupName == "" {
+			testGroupName = j.Name
+		}
+		tabName = testGroupName
+		if c.config.FindTestGroup(testGroupName) != nil {
+			if mustMakeGroup {
+				return fmt.Errorf("test group %q already exists", testGroupName)
+			}
+		} else {
+			var prefix string
+			if j.DecorationConfig != nil && j.DecorationConfig.GCSConfiguration != nil {
+				prefix = path.Join(j.DecorationConfig.GCSConfiguration.Bucket, j.DecorationConfig.GCSConfiguration.PathPrefix)
+			} else if pc.Plank.DefaultDecorationConfig != nil && pc.Plank.DefaultDecorationConfig.GCSConfiguration != nil {
+				prefix = path.Join(pc.Plank.DefaultDecorationConfig.GCSConfiguration.Bucket, j.DecorationConfig.GCSConfiguration.PathPrefix)
+			} else {
+				return fmt.Errorf("job %s: couldn't figure out a default decoration config", j.Name)
+			}
+
+			g := &config.TestGroup{
+				Name:      testGroupName,
+				GcsPrefix: path.Join(prefix, prowGCS.RootForSpec(&downwardapi.JobSpec{Job: j.Name, Type: jobType})),
```

Comment:
Ooh, might be nice to have a default description noting the job name this was constructed from? (The config yaml would be nicer but I don't know that that's plausible).

Reply:
Done.

Reply:
Sweet!